### DISCUSSION TITLE
`Kyuubi.session.idle.timeout` doesn't work

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#5060] [Q&A] [answered]
I created Kyuubi session with these configs:
```
      "kyuubi.session.engine.idle.timeout": "PT2M",
      "kyuubi.session.idle.timeout": "PT1M"
```

But after several minutes, the session is still live here. The engine closed after 15 mins and session was still here.

**Comment by pan3793** [comment]
> But after several minutes, the session is still live here.

It sounds like an unexpected behavior, any clues in the logs?

> The engine closed after 15 mins and session was still here.

It was enhanced by https://github.com/apache/kyuubi/issues/4847.

**Reply by HaoYang670** [reply]
No, there is no log about the Kyuubi session. Could you reproduce the problem?

**Reply by tgravescs** [reply]
It looks like the problem might be that _lastIdleTime in the Session is set to 0.  Here we create a session, which creates and engine.  The first session idle times out because nothing is run, we create a second session attaching to the engine and then run an operation. At that point it seems stuck with that _lastIdleTime = 0 so the sessionmanager never times it out:

```
 if (session.lastAccessTime + session.sessionIdleTimeoutThreshold <= current &&
              session.getNoOperationTime > session.sessionIdleTimeoutThreshold) 
```

I haven't figured out why that is 0 though when you actually run a statement.  Maybe its not recognizing its done?

**Reply by tgravescs** [reply]
oh in my case I'm running GROUP share level

**Reply by tgravescs** [reply]
I'm guessing the problem is the operation isn't being closed??

**Reply by tgravescs** [reply]
tried closing but that didn't help. I set the operation.idle.timeout and then things worked after that but I would think you wouldn't have to wait for them.  But maybe there is something around that I'm missing.

**Comment by pan3793** [comment, marked as answer]
> I created Kyuubi session with these configs: ...

I suppose `kyuubi.session.idle.timeout` does not work on the session level now, have you tried to configure it in the `kyuubi-defaults.conf`?

https://github.com/apache/kyuubi/blob/v1.7.1/kyuubi-common/src/main/scala/org/apache/kyuubi/session/AbstractSession.scala#L56

**Reply by HaoYang670** [reply]
Thanks. I move the session and engine timeout settings to kyuubi-defualt and it works.

**Reply by HaoYang670** [reply]
But the operation timeout still doesn't work. And it blocks the session to be removed

**Reply by HaoYang670** [reply]
> But the operation timeout still doesn't work. And it blocks the session to be removed

Ok, it finally works. I guess I miss setting the interval. But I don't find a config like `operation.timeout.interval`

**Reply by pan3793** [reply]
> But I don't find a config like `operation.timeout.interval`

It's controlled by `kyuubi.session.check.interval`, it seems the description is misleading. should be 
```patch
  val SESSION_CHECK_INTERVAL: ConfigEntry[Long] = buildConf("kyuubi.session.check.interval")
-   .doc("The check interval for session timeout.")
+   .doc("The check interval for session/operation timeout.")
```

And I think we can improve the `kyuubi.session.idle.timeout` to make it configurable on the client side at the session level, which makes the behavior more intuitive.

**Reply by HaoYang670** [reply]
> And I think we can improve the kyuubi.session.idle.timeout to make it configurable on the client side at the session level, which makes the behavior more intuitive

Yeh, it could be better.

**Reply by HaoYang670** [reply]
Filed https://github.com/apache/kyuubi/issues/5098
